connector/ldap/ldap.go

```diff
@@ -496,52 +503,107 @@ func (c *ldapConnector) groups(ctx context.Context, user ldap.Entry) ([]string,
+	for {
+		// Temporal variable used to reset variable groups on each cycle
+		var nextLevelGroups []*ldap.Entry
+		for _, group := range groups {
+
+			name := getAttr(*group, c.GroupSearch.NameAttr)
+			if name == "" {
+				// Be obnoxious about missing missing attributes. If the group entry is
+				// missing its name attribute, that indicates a misconfiguration.
+				//
+				// In the future we can add configuration options to just log these errors.
+				return nil, fmt.Errorf("ldap: group entity %q missing required attribute %q",
+					group.DN, c.GroupSearch.NameAttr)
+			}
 
-		gotGroups := false
-		if err := c.do(ctx, func(conn *ldap.Conn) error {
-			c.logger.Infof("performing ldap search %s %s %s",
-				req.BaseDN, scopeString(req.Scope), req.Filter)
-			resp, err := conn.Search(req)
-			if err != nil {
-				return fmt.Errorf("ldap: search failed: %v", err)
+			// Prevent duplicates and circular hierarchy
+			exit := false
+			for _, groupName := range groupNames {
+				if name == groupName {
+					c.logger.Infof("Found duplicate group with name %s", name)
+					exit = true
+					break
+				}
+				c.logger.Infof("Comparing %s with %s", name, groupName)
+			}
+			if exit {
+				continue
+			}
+
+			groupNames = append(groupNames, name)
+			if c.GroupSearch.Inheritance {
+				if obtainedGroups, _, err := c.queryGroups(ctx, group.DN); err == nil {
+					// Keep searching upwards
+					if len(obtainedGroups) != 0 {
+						nextLevelGroups = append(nextLevelGroups, obtainedGroups...)
+					} else {
+						c.logger.Infof("Didn't find parents for group with name %s", name)
+					}
+				} else {
+					return nil, err
+				}
 			}
-			gotGroups = len(resp.Entries) != 0
-			groups = append(groups, resp.Entries...)
-			return nil
-		}); err != nil {
-			return nil, err
 		}
-		if !gotGroups {
-			// TODO(ericchiang): Is this going to spam the logs?
-			c.logger.Errorf("ldap: groups search with filter %q returned no groups", filter)
+		// If there's no remaining levels -> exit loop
+		if len(nextLevelGroups) == 0 {
+			break
 		}
+		// reassign groups for next loop
+		groups = nextLevelGroups
+
 	}
+	return groupNames, nil
 
-	var groupNames []string
-	for _, group := range groups {
-		name := getAttr(*group, c.GroupSearch.NameAttr)
-		if name == "" {
-			// Be obnoxious about missing missing attributes. If the group entry is
-			// missing its name attribute, that indicates a misconfiguration.
-			//
-			// In the future we can add configuration options to just log these errors.
-			return nil, fmt.Errorf("ldap: group entity %q missing required attribute %q",
-				group.DN, c.GroupSearch.NameAttr)
-		}
+}
 
-		groupNames = append(groupNames, name)
+//  Abrstraction query of Groups and reuse code
+func (c *ldapConnector) queryGroups(ctx context.Context, dn string) ([]*ldap.Entry, string, error) {
+	var groups []*ldap.Entry
+	filter := fmt.Sprintf("(%s=%s)", c.GroupSearch.GroupAttr, ldap.EscapeFilter(dn))
```

Comment:
I don't think we can assume that the attribute used for determining membership of a user will be the same for determining the membership of the group.

Reply:
Usually, on ActiveDirectory GroupAttr is  "member". 
I find it wrong to have a LDAP setup where for example I have the attribute "member" for users and some other word for groups. It can happen of course but I find more confusing for most users to have two different attributes being the same most of the cases. 

For example:
It can happen that someone instead of assigning 
```yaml
groupAttr:"member"
recursionAttr:"member"
```
assigns
```yaml
groupAttr:"member"
recursionAttr:"memberOf"
``` 
by thinking the attributes make different things. But is up to you to make the decision.
What do you prefer?